code/arduino/motors/regulation.h

```diff
@@ -0,0 +1,80 @@
+#ifndef REGULATION_H
+#define REGULATION_H
+
+#include "motor.h"
+
+// Abstract class. Should not be instantiated.
+class Regulation {
+    public:
+        Regulation(Motor *left, Motor *right);
+        void set_setpoint(int setpoint);
+        void tune();
+        void set_max_speed(int maxspeed);
+        void resume();
+        void stop();
+        bool is_stopped() const;
+        bool is_finished() const;
+
+    protected:
+        int setpoint;
+
+        virtual int get_maxspeed();
+        virtual float get_rotation_error();
+        virtual float get_lead_error();
+
+    private:
+        // Regulation gain.
+        const float KP_LEAD = 0.6;
+        const float KP_ROT = 0.6;
+        const float KI_LEAD = 0.006;
+        const float KI_ROT = 0.01;
+
+        const int SUM_ERRORS_LIMIT = 300;
+        const int INTEGRAL_SATURATION = 60;
+        // Limit the speed command to the last command + PROGRESSIVE_COMMAND
+        // to avoid demanding too much to the motors and the alim.
+        const int PROGRESSIVE_COMMAND = 10;
+        // Difference of commands to decide if we need to increase the max
+        // speed of a motor.
+        const int COMMAND_DELTA = 3;
+        const int MAX_SPEED_BOOST = 5;
+
+        const int REGULATION_PRECISION = 2;
+
+        const int MAXSPEED = 50;
```

Comment:
Tu peux passer toute les constantes en static
```c++
static constexpr float KP_LEAD = 0.6;
static constexpr float KP_ROT = 0.6;
static constexpr float KI_LEAD = 0.006;
static constexpr float KI_ROT = 0.01;

static const int SUM_ERRORS_LIMIT = 300;
static const int INTEGRAL_SATURATION = 60;
// Limit the speed command to the last command + PROGRESSIVE_COMMAND
// to avoid demanding too much to the motors and the alim.
 static const int PROGRESSIVE_COMMAND = 10;
// Difference of commands to decide if we need to increase the max
// speed of a motor.
static const int COMMAND_DELTA = 3;
static const int MAX_SPEED_BOOST = 5;
static const int REGULATION_PRECISION = 2;
static const int MAXSPEED = 50;
```

Le compilateur ne va même pas allouer de stockage pour la variable et simplement remplacer le nom par la valeur où elle est utilisé. Sans le static, le compilateur ne peut pas supposer qu'elle n'est pas utilisé ailleurs et ne peut pas faire cette modification.